### PR TITLE
Non-record: TTT + QAT on Consumer GPU (val_bpb=1.5382)

### PR DESCRIPTION
**Pull request by Dannybc123**
## val_bpb = 1.5382 (single run, standard eval + TTT)

Non-record: local consumer GPU only (1x RTX 5070 Ti, 12GB VRAM). Artifact: 11.5 MB. 300s training + 619s eval (includes TTT).

This submission was developed entirely on a laptop GPU using AI-assisted autonomous experimentation. No H100 access was used during development.

---

### Key Finding: QAT Makes Training Faster While Improving Quality

Adding fake quantization noise during training (QAT) made the model train **faster** (362ms/step vs 399ms/step) while producing a better post-quantization model. The `torch.round` + `clamp` ops in the straight-through estimator are cheaper than full-precision matmuls, so QAT is effectively free compute.

Combined with test-time training (TTT) on both attention and MLP output projections, this produced a **1.5382 BPB** result (vs 1.5540 baseline), a 1.0% improvement.

### Methodology: AI-Assisted Autonomous Experimentation

Developed using an autoresearch loop — an AI agent (Claude) autonomously modifies `train_gpt.py`, runs 5-minute experiments on a local RTX 5070 Ti, evaluates results, and iterates. 15 experiments were run with strict one-variable-at-a-time methodology.

### Architecture

9 layers, dim=512, 8 heads, 4 KV heads. MLP 2x (hidden=1024). Tied embeddings. Stock baseline architecture — no architectural changes.

### Key Modifications

- **QAT in `CastedLinear.forward()`:** Simulate int8 quantization during training via straight-through estimator. 7 lines of code.
- **TTT in `ttt_eval_val()`:** 1 SGD step (LR=1e-4) on both `attn.proj.weight` and `mlp.proj.weight` per eval batch. ~30 lines of code.

### Results

| Metric | Value |
|--------|-------|
| val_bpb (with TTT) | **1.5382** |
| val_bpb (base, no TTT) | 1.5435 |
| TTT delta | -0.005 |
| Artifact | 11,506,921 bytes |
| Steps | 779 |
| ms/step | 385 |
| Hardware | 1x RTX 5070 Ti (12GB) |

### Experiment History (15 experiments)

| # | Experiment | BPB | Status |
|---|-----------|-----|--------|
| E001 | Baseline | 1.5540 | control |
| E003 | TTT on MLP proj | 1.5265 | keep |
| E009 | TTT both attn+MLP | 1.5462 | keep |
| E011 | QAT (fake int8) | 1.5290 | keep |
| E012 | TopK 50% sparsity | 1.5566 | discard |
| E013 | MLP_MULT=3 | 1.5318 | discard |

### What Didn't Work

- **TopK MLP sparsity:** Hurt base model — not enough capacity at dim=512 for forced sparsity
- **Multi-step TTT (2 steps):** No improvement, 17min eval
- **Wider MLP (3x):** Bigger != better at ~750 training steps
- **TTT LR=3e-4:** Overshoots adaptation

### Acknowledgments

Built on the naive baseline. TTT inspired by PR #77 (samacqua). Autoresearch methodology adapted from [Karpathy's autoresearch](https://github.com/karpathy/autoresearch).

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: TTT + QAT on Consumer GPU (val_bpb=1.5382)

**Compliance flag: Pre-Quant TTT violation (adapt-then-score)**


## PR #263 — Non-record: TTT + QAT on Consumer GPU (val_bpb=1.5382)

**Head SHA:** `046b7d2fdcb161ab7e4c082057fc7296ba01df62`
**Primary file:** `records/track_non_record_16mb/2026-03-20_TTT-QAT-LocalGPU/train_gpt.py`

---

### Check 1: N-gram family bug (CLOSE trigger)
No n-gram, BigramHash, or hash-based lookup structures found. No target token used as a hash key. **CLEAN.**

### Check 2: Pre-Quant TTT (CLOSE trigger)
The `ttt_eval_val()` function (lines 335–426) implements **adapt-then-score**, not score-first-per-chunk:

```python
# --- TTT: adapt W_down to this batch ---
for _ in range(ttt_steps):
    adapt_loss = base_model(x, y)          # forward on (x, y)
    grads = torch.autograd.grad(adapt_loss, ttt_params)
    p.sub_(g, alpha=ttt_lr)                # gradient step

# --- Score with adapted weights ---
batch_loss = base_model(x, y).detach()    # SAME (x, y) scored after adaptation
```

The model adapts its weights using the target sequence `y` before scoring that same sequence. This gives the model lookahead into the targets prior to producing the loss that counts toward BPB. The optimizer is SGD (not AdamW) with a default of 1 step (configurable via `TTT_STEPS`), but the structural violation is identical to the Pre-Quant TTT CLOSE pattern: **targets are leaked into the model state before scoring**.

The task spec requires **score-first-per-chunk** for legal TTT: forward pass → record loss → then optionally adapt. This PR inverts that order.

**CLOSE — Pre-Quant TTT violation (adapt-before-score on same batch).**

The TTT implementation in `ttt_eval_val()` adapts model weights on a batch `(x, y)` and then immediately scores the **same batch** with the adapted weights. This means the targets `y` have been leaked into the model parameters before the BPB-counting forward pass. This violates the score-first-per-chunk requirement established in PR #1413 and the rulings in Issue #677.

The legal pattern requires scoring each chunk under `torch.no_grad()` **before** any gradient step — here, the gradient step comes first.

**Verdict:** CLOSE — Pre-Quant TTT violation (adapt-then-score on same batch).

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** This PR's TTT implementation adapts on the scored batch before scoring it, which violates score-first discipline. Recommend **CLOSE** unless the author restructures to score-first-per-chunk (PR #1413 pattern).

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
